src/bespokelabs/curator/request_processor/online_request_processor.py

```diff
@@ -0,0 +1,312 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+import datetime
+import time
+from typing import Optional
+from tqdm import tqdm
+import logging
+import asyncio
+import aiohttp
+import os
+import json
+import resource
+
+from bespokelabs.curator.dataset import Dataset
+from bespokelabs.curator.request_processor.base_request_processor import BaseRequestProcessor
+from bespokelabs.curator.prompter.prompter import PromptFormatter
+from bespokelabs.curator.request_processor.generic_request import GenericRequest
+from bespokelabs.curator.request_processor.event_loop import run_in_event_loop
+from bespokelabs.curator.request_processor.generic_response import GenericResponse
+import aiofiles
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class StatusTracker:
+    """Tracks the status of all requests."""
+
+    num_tasks_started: int = 0
+    num_tasks_in_progress: int = 0
+    num_tasks_succeeded: int = 0
+    num_tasks_failed: int = 0
+    num_tasks_already_completed: int = 0
+    num_api_errors: int = 0
+    num_other_errors: int = 0
+    available_request_capacity: float = 0
+    available_token_capacity: float = 0
+    last_update_time: float = field(default_factory=time.time)
+    max_requests_per_minute: int = 0
+    max_tokens_per_minute: int = 0
+    pbar: tqdm = field(default=None)
+    response_cost: float = 0
+
+    def __str__(self):
+        return (
+            f"Tasks - Started: {self.num_tasks_started}, "
+            f"In Progress: {self.num_tasks_in_progress}, "
+            f"Succeeded: {self.num_tasks_succeeded}, "
+            f"Failed: {self.num_tasks_failed}, "
+            f"Already Completed: {self.num_tasks_already_completed}\n"
+            f"Errors: {self.num_other_errors}"
+        )
+
+    def update_capacity(self):
+        """Update available capacity based on time elapsed"""
+        current_time = time.time()
+        seconds_since_update = current_time - self.last_update_time
+
+        self.available_request_capacity = min(
+            self.available_request_capacity
```

Comment:
specifically for openAI they have a new header which gives you this https://platform.openai.com/docs/guides/rate-limits#rate-limits-in-headers
x-ratelimit-remaining-requests
x-ratelimit-remaining-tokens

Reply:
The current implementation is fine for now. We can support this later.